File: JsonParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

class JsonParser
{
    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: JsonParser <file.json> [flatten|structure|pretty]");
            Console.WriteLine("  flatten   - print all values with their key paths (default)");
            Console.WriteLine("  structure - print nested structure overview");
            Console.WriteLine("  pretty    - pretty-print the JSON");
            return;
        }

        string filepath = args[0];
        string mode = args.Length > 1 ? args[1] : "flatten";

        string jsonText = File.ReadAllText(filepath);
        using JsonDocument doc = JsonDocument.Parse(jsonText);
        JsonElement root = doc.RootElement;

        switch (mode)
        {
            case "flatten":
                foreach (var (path, value) in Flatten(root, ""))
                    Console.WriteLine($"{path} = {value}");
                break;

            case "structure":
                PrintStructure(root, 0);
                break;

            case "pretty":
                var options = new JsonSerializerOptions { WriteIndented = true };
                Console.WriteLine(JsonSerializer.Serialize(JsonSerializer.Deserialize<object>(jsonText), options));
                break;

            default:
                Console.WriteLine($"Unknown mode: {mode}");
                break;
        }
    }

    static IEnumerable<(string Path, string Value)> Flatten(JsonElement element, string prefix)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Array:
                int i = 0;
                foreach (var item in element.EnumerateArray())
                    foreach (var entry in Flatten(item, $"{prefix}[{i++}]"))
                        yield return entry;
                break;

            case JsonValueKind.Object:
                foreach (var prop in element.EnumerateObject())
                {
                    string newKey = string.IsNullOrEmpty(prefix) ? prop.Name : $"{prefix}.{prop.Name}";
                    foreach (var entry in Flatten(prop.Value, newKey))
                        yield return entry;
                }
                break;

            default:
                yield return (prefix, element.ToString());
                break;
        }
    }

    static void PrintStructure(JsonElement element, int indent)
    {
        string pad = new string(' ', indent * 2);

        switch (element.ValueKind)
        {
            case JsonValueKind.Array:
                int count = 0;
                foreach (var _ in element.EnumerateArray()) count++;
                Console.WriteLine($"{pad}Array ({count} items)");
                int i = 0;
                foreach (var item in element.EnumerateArray())
                {
                    Console.WriteLine($"{pad}  [{i++}]:");
                    PrintStructure(item, indent + 2);
                }
                break;

            case JsonValueKind.Object:
                int keyCount = 0;
                foreach (var _ in element.EnumerateObject()) keyCount++;
                Console.WriteLine($"{pad}Object ({keyCount} keys)");
                foreach (var prop in element.EnumerateObject())
                {
                    Console.WriteLine($"{pad}  {prop.Name}:");
                    PrintStructure(prop.Value, indent + 2);
                }
                break;

            default:
                Console.WriteLine($"{pad}{element}");
                break;
        }
    }
}

File: json_parser.py
#!/usr/bin/env python3

import json
import sys


def flatten(data, prefix=""):
    """Recursively flatten nested arrays/objects with dotted key paths."""
    results = []
    if isinstance(data, list):
        for i, item in enumerate(data):
            results.extend(flatten(item, f"{prefix}[{i}]"))
    elif isinstance(data, dict):
        for key, value in data.items():
            new_key = f"{prefix}.{key}" if prefix else key
            results.extend(flatten(value, new_key))
    else:
        results.append((prefix, data))
    return results


def print_structure(data, indent=0):
    """Pretty-print nested structure with indentation."""
    pad = "  " * indent
    if isinstance(data, list):
        print(f"{pad}Array ({len(data)} items)")
        for i, item in enumerate(data):
            print(f"{pad}  [{i}]:")
            print_structure(item, indent + 2)
    elif isinstance(data, dict):
        print(f"{pad}Object ({len(data)} keys)")
        for key, value in data.items():
            print(f"{pad}  {key}:")
            print_structure(value, indent + 2)
    else:
        print(f"{pad}{repr(data)}")


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 json_parser.py <file.json> [flatten|structure|pretty]")
        print("  flatten   - print all values with their key paths (default)")
        print("  structure - print nested structure overview")
        print("  pretty    - pretty-print the JSON")
        sys.exit(1)

    filepath = sys.argv[1]
    mode = sys.argv[2] if len(sys.argv) > 2 else "flatten"

    with open(filepath) as f:
        data = json.load(f)

    if mode == "flatten":
        for path, value in flatten(data):
            print(f"{path} = {repr(value)}")
    elif mode == "structure":
        print_structure(data)
    elif mode == "pretty":
        print(json.dumps(data, indent=2))
    else:
        print(f"Unknown mode: {mode}")
        sys.exit(1)


if __name__ == "__main__":
    main()
